auth-foundation/src/main/java/com/okta/authfoundation/credential/storage/TokenEntity.kt

```diff
@@ -0,0 +1,71 @@
+/*
+ * Copyright 2024-Present Okta, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.okta.authfoundation.credential.storage
+
+import androidx.room.Entity
+import androidx.room.PrimaryKey
+import androidx.room.TypeConverters
+import com.okta.authfoundation.credential.storage.typeconverters.JsonObjectTypeConverter
+import com.okta.authfoundation.credential.storage.typeconverters.StringStringMapTypeConverter
+import kotlinx.serialization.json.JsonObject
+
+@Entity
+@TypeConverters(
+    StringStringMapTypeConverter::class,
+    JsonObjectTypeConverter::class
+)
+internal data class TokenEntity(
+    @PrimaryKey
+    val id: String,
+    val encryptedToken: ByteArray,
+    val tags: Map<String, String>,
+    val payloadData: JsonObject,
+    val keyAlias: String,
+    val userAuthenticationRequired: Boolean,
```

Comment:
What do `keyAlias` and `userAuthenticationRequired` do?

Reply:
Good question. keyAlias is the RSA key that's used for encrypting the token. userAuthenticationRequired means that the user needs to do Biometric or PIN based authentication, in which case we pop up a Biometric prompt.

Reply:
I see.  On Android, or ChromeOS, is there any other kind of user authentication that's possible?  For example, on Apple platforms, you can constrain it to pin, generic biometrics, touchID, faceID, etc.

What I'm getting at is, should this be an enum, or a simple boolean?

Reply:
Good question. I think I should change this to an enum in case I want to extend the functionality to newer authentication methods in the future. And that also makes decryption a bit more user friendly too.

On the actual Android APIs side, things are a bit messy for Android 10 and below. On Android 11+, the users can specify exactly which combination of methods they want through flags. But, on Android 10 and below, users have to use specific timeouts to choose authentication methods: 0 for Bio + Pin, and -1 for Bio only.

Either way, I think enum should make things a bit simpler to extend in the future. So, I will use that instead.

Reply:
🤔 With platform version differences, and different capabilities for API revisions, it feels like this part needs to be abstracted away anyway.  Using an enum will certainly help, but perhaps this is an area where we can future-proof the design.

For example, the enum values could be tagged as only supporting certain API versions so the developer gets compiler errors if they use unsupported enum types in their apps.  Or some sort of fallback behavior can be built in to map the behavior for older platforms.

This is similar to the Swift SDK's WebAuthenticationUI handling.  Different iOS versions have different SDKs and classes that handle web-based authentication.  e.g. iOS 12 uses ASWebAuthenticationSession, iOS 11 SFAuthenticationSession, and prior to that plain SFSafariBrowser.  So I created a [WebAuthenticationProvider protocol](https://github.com/okta/okta-mobile-swift/blob/master/Sources/WebAuthenticationUI/Providers/WebAuthenticationProvider.swift) that defined the interface for platform-specific capabilities, and then wrote individual classes that worked on those different platform versions.

That might be overkill for this use-case, but either way, abstracting away the platform- or version-specific handling for something that changes based on API version or environment is useful, not only to deal with the current known conditions, but to future-proof your implementation to support other platforms like ChromeOS or anything else we might nee to support later.

Reply:
For now, I don't think it's necessary to have platform-specific checks. Android supports either Bio or Pin for authentication. Bio only, or Bio + Pin are supported on all platforms, and I don't see much value in having a Pin only authentication. But, the door for that is open with the enum approach now